File: Python/501. Find Mode in Binary Search Tree.py
'''
1.递归
2.迭代
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def findMode(self, root: Optional[TreeNode]) -> List[int]:
#         res = []
#         pre = None
#         cur_count = 1
#         max_count = 1
#         def traversal(node):
#             nonlocal pre, cur_count, max_count, res
#             if not node:
#                 return
#             traversal(node.left)
#             if pre:
#                 if pre.val == node.val:
#                     cur_count += 1
#                 else:
#                     cur_count = 1
#             pre = node
#             if cur_count == max_count:
#                 res.append(node.val)
#             elif cur_count > max_count:
#                 max_count = cur_count
#                 res = [node.val]
#             traversal(node.right)
#         traversal(root)
#         return res

class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        pre = None
        cur_count = 1
        max_count = 1
        st = [root]
        while st:
            cur = st.pop(-1)
            if cur:
                if cur.right: st.append(cur.right)
                st.append(cur)
                st.append(None)
                if cur.left: st.append(cur.left)
            else:
                cur = st.pop(-1)
                if pre:
                    if pre.val == cur.val: cur_count += 1
                    else: cur_count = 1
                pre = cur
                if cur_count == max_count: res.append(cur.val)
                elif cur_count > max_count:
                    max_count = cur_count
                    res = [cur.val]
        return res

File: C++/501. Find Mode in Binary Search Tree.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 // class Solution {
 // public:
 //     int max_count = 1;
 //     int cur_count = 1;
 //     vector<int> res;
 //     TreeNode* pre = NULL;
 //     void traversal(TreeNode* cur){
 //         if (cur == NULL) return;
 //         if (cur->left) traversal(cur->left);
 //         if (pre){
 //             if (pre->val == cur->val) cur_count++;
 //             else cur_count = 1;
 //         }
 //         pre = cur;
 //         if (cur_count == max_count){
 //             res.push_back(cur->val);
 //         }else if (cur_count > max_count){
 //             res.clear();
 //             res.push_back(cur->val);
 //             max_count = cur_count;
 //         }
 //         if (cur->right) traversal(cur->right);
 //     }
 //     vector<int> findMode(TreeNode* root) {
 //         res.clear();
 //         traversal(root);
 //         return res;
 //     }
 // };

class Solution {
public:
    vector<int> findMode(TreeNode* root) {
        stack<TreeNode*> st;
        int max_count = 1;
        int cur_count = 1;
        vector<int> res;
        TreeNode* pre = NULL;
        st.push(root);
        while (!st.empty()) {
            TreeNode* cur = st.top(); st.pop();
            if (cur) {
                if (cur->right) st.push(cur->right);
                st.push(cur);
                st.push(NULL);
                if (cur->left) st.push(cur->left);
            }
            else {
                TreeNode* cur = st.top(); st.pop();
                if (pre) {
                    if (pre->val == cur->val) cur_count++;
                    else cur_count = 1;
                }
                pre = cur;
                if (cur_count == max_count) res.push_back(cur->val);
                else if (cur_count > max_count) {
                    max_count = cur_count;
                    res.clear();
                    res.push_back(cur->val);
                }
            }
        }
        return res;
    }
};